base/intfuncs.jl

```diff
@@ -48,33 +48,46 @@ function gcd(a::T, b::T) where T<:Integer
 end
 
 function gcd(a::T, b::T) where T<:BitInteger
-    a == 0 && return checked_abs(b)
-    b == 0 && return checked_abs(a)
-    r = _gcd(a, b)
-    signbit(r) && __throw_gcd_overflow(a, b)
-    return r
+    a == 0 && return Base.checked_abs(b)
+    b == 0 && return Base.checked_abs(a)
+    if a isa Signed && a == typemin(T)
+        if a == b
+            Base.__throw_gcd_overflow(a, b)
+        else
+            a, b = b, a
+        end
+    end
+    return _gcd(a, b)
 end
 @noinline __throw_gcd_overflow(a, b) =
     throw(OverflowError(LazyString("gcd(", a, ", ", b, ") overflows")))
 
+function absdiff(x::T,y::T) where {T<:Unsigned}
+    d = max(x,y) - min(x,y)
```

Comment:
I've tried making this faster by avoiding a branch of either `min` or `max` like this:

```
function ad(x::T, y::T) where T <: Unsigned
    d,f =  Base.Checked.sub_with_overflow(x,y)
    res = (f*(-2 % T)) ⊻ d
    res, res
end
```

which produces very nice assembly without branching on its own, but then doesn't want to vectorize if it's put in a loop.. I think this is due to `sub_with_overflow` not being able to vectorize, even though the used LLVM intrinsic `@llvm.usub.with.overflow` [has vectorized versions available as well](https://llvm.org/docs/LangRef.html#llvm-usub-with-overflow-intrinsics). 

I'm not sure if or how julia could vectorize this, since it seems very tricky to replace one intrinsic for another here.. Though I'd be very interested in seeing if this could be faster than a vectorized min/max, due to a reduced need for branching!

Reply:
Many operations on gpr set flags when they overflow, which is why it's efficient.
Vector instructions do not set flags, so you have to check for overflow manually.
This paper discusses an approach: https://dl.acm.org/doi/pdf/10.1145/3485539

Probably the easier way to generate SIMD code for this is via cmp and blend.

Anyway, `gcd` isn't going to be auto vectorizing any time soon. You can vectorize it:
https://github.com/JuliaSIMD/VectorizationBase.jl/blob/f1bac4cfb10980cb6dd5441d80c94681087263cd/src/special/misc.jl#L160-L184
but I imagine that it
1. won't normally be inlining
2. if it does, LLVM won't know how to vectorize a `while` loop.

Reply:
Well, your `max(x,y) - min(x,y)` approach does actually vectorize quite nicely once its put into a loop. However, due to those `min`/`max`, it is forced to branch vectorize as well. I had hoped I could do a similar trick for that version above, but it doesn't want to. I also found [this](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=sbb&ig_expand=6055) instruction/intrinsic, but I'm not 100% sure if that's related/usable. It at least seems to kind of be what would be interesting here, if only for Int32?

<details>
<summary>If you're interested, here's my code.</summary>

```julia
julia> function check(f, data)
           s = zero(eltype(data))
           @inbounds for idx in axes(data, 1)
               s += first(f(data[idx, 1], data[idx, 2]))
           end
           s
       end
check (generic function with 1 method)

julia> function absdiff(x::T,y::T) where {T<:Unsigned}
           d = max(x,y) - min(x,y)
           d, d
       end
absdiff (generic function with 1 method)

julia> function ad(x::T,y::T) where {T<:Unsigned}
           d,f = Base.Checked.sub_with_overflow(x, y)
           res = (f*(-2 % T)) ⊻ d
           res, res
       end
ad (generic function with 1 method)

julia> data = rand(UInt16, 10000000, 2);
```

which is then called

```julia
julia> @code_native check(ad, data)
julia> @code_native check(absdiff, data)
```

respectively.

I get these:

```asm
julia> @code_native debuginfo=:none check(ad, data)
	.text
	.file	"check"
	.globl	julia_check_131                 # -- Begin function julia_check_131
	.p2align	4, 0x90
	.type	julia_check_131,@function
julia_check_131:                        # @julia_check_131
	.cfi_startproc
# %bb.0:                                # %top
	movq	24(%rdi), %r9
	testq	%r9, %r9
	je	.LBB0_1
# %bb.2:                                # %L15.preheader
	movq	(%rdi), %rdx
	leaq	(%rdx,%r9,2), %r8
	xorl	%edi, %edi
	xorl	%eax, %eax
	.p2align	4, 0x90
.LBB0_3:                                # %L15
                                        # =>This Inner Loop Header: Depth=1
	movzwl	(%rdx,%rdi,2), %esi
	subw	(%r8,%rdi,2), %si
	sbbl	%ecx, %ecx
	addl	%ecx, %ecx
	xorl	%esi, %ecx
	addl	%ecx, %eax
	incq	%rdi
	cmpq	%rdi, %r9
	jne	.LBB0_3
# %bb.4:                                # %L40
                                        # kill: def $ax killed $ax killed $eax
	retq
.LBB0_1:
	xorl	%eax, %eax
                                        # kill: def $ax killed $ax killed $eax
	retq
.Lfunc_end0:
	.size	julia_check_131, .Lfunc_end0-julia_check_131
	.cfi_endproc
                                        # -- End function
	.section	".note.GNU-stack","",@progbits
```

```asm
julia> @code_native debuginfo=:none check(absdiff, data)
	.text
	.file	"check"
	.globl	julia_check_162                 # -- Begin function julia_check_162
	.p2align	4, 0x90
	.type	julia_check_162,@function
julia_check_162:                        # @julia_check_162
	.cfi_startproc
# %bb.0:                                # %top
	movq	24(%rdi), %r8
	testq	%r8, %r8
	je	.LBB0_1
# %bb.2:                                # %L15.preheader
	movq	(%rdi), %r9
	cmpq	$16, %r8
	jae	.LBB0_4
# %bb.3:
	movl	$1, %esi
	xorl	%eax, %eax
	jmp	.LBB0_7
.LBB0_1:
	xorl	%eax, %eax
                                        # kill: def $ax killed $ax killed $eax
	retq
.LBB0_4:                                # %vector.ph
	movq	%r8, %rdi
	andq	$-16, %rdi
	leaq	1(%rdi), %rsi
	leaq	(%r9,%r8,2), %rax
	vpxor	%xmm0, %xmm0, %xmm0
	xorl	%ecx, %ecx
	.p2align	4, 0x90
.LBB0_5:                                # %vector.body
                                        # =>This Inner Loop Header: Depth=1
	vmovdqu	(%r9,%rcx,2), %ymm1
	vmovdqu	(%rax,%rcx,2), %ymm2
	vpmaxuw	%ymm1, %ymm2, %ymm3
	vpminuw	%ymm1, %ymm2, %ymm1
	vpsubw	%ymm1, %ymm3, %ymm1
	vpaddw	%ymm0, %ymm1, %ymm0
	addq	$16, %rcx
	cmpq	%rcx, %rdi
	jne	.LBB0_5
# %bb.6:                                # %middle.block
	vextracti128	$1, %ymm0, %xmm1
	vpaddw	%xmm1, %xmm0, %xmm0
	vpshufd	$238, %xmm0, %xmm1              # xmm1 = xmm0[2,3,2,3]
	vpaddw	%xmm1, %xmm0, %xmm0
	vpshufd	$85, %xmm0, %xmm1               # xmm1 = xmm0[1,1,1,1]
	vpaddw	%xmm1, %xmm0, %xmm0
	vpsrld	$16, %xmm0, %xmm1
	vpaddw	%xmm1, %xmm0, %xmm0
	vmovd	%xmm0, %eax
	cmpq	%rdi, %r8
	je	.LBB0_9
.LBB0_7:                                # %L15.preheader16
	leaq	(%r9,%r8,2), %r10
	decq	%rsi
	.p2align	4, 0x90
.LBB0_8:                                # %L15
                                        # =>This Inner Loop Header: Depth=1
	movzwl	(%r9,%rsi,2), %ecx
	movzwl	(%r10,%rsi,2), %edi
	cmpw	%cx, %di
	movl	%edi, %edx
	cmovbl	%ecx, %edx
	cmovbl	%edi, %ecx
	subl	%ecx, %edx
	addl	%edx, %eax
	incq	%rsi
	cmpq	%rsi, %r8
	jne	.LBB0_8
.LBB0_9:                                # %L40
                                        # kill: def $ax killed $ax killed $eax
	vzeroupper
	retq
.Lfunc_end0:
	.size	julia_check_162, .Lfunc_end0-julia_check_162
	.cfi_endproc
                                        # -- End function
	.section	".note.GNU-stack","",@progbits
```
</details>

Reply:
It won't vectorize from inside the `while a != 0` loop.